controllers/operator/openstack_controller.go

```diff
@@ -372,6 +381,182 @@ func (r *OpenStackReconciler) renderAndApply(
+	subscriptionList.SetGroupVersionKind(subscriptionGVR.GroupVersion().WithKind("Subscription"))
+	err = r.Client.List(ctx, subscriptionList, &client.ListOptions{Namespace: instance.Namespace})
+	if err != nil {
+		return err
+	}
+	for _, subscription := range subscriptionList.Items {
+		Log.Info("Found Subscription", "name", subscription.GetName())
+		if isServiceOperatorResource(subscription.GetName()) {
+			err = r.Client.Delete(ctx, &subscription)
+			if err != nil {
+				return err
+			}
+			Log.Info("Subscription deleted successfully", "name", subscription.GetName())
+		}
+	}
+
+	// lookup the installplan which has the clusterServiceVersionNames we removed above
+	// there will be just a single installPlan that has all of them referenced
+	installPlanList := &uns.UnstructuredList{}
+	installPlanList.SetGroupVersionKind(installPlanGVR.GroupVersion().WithKind("InstallPlan"))
+
+	err = r.Client.List(ctx, installPlanList, &client.ListOptions{Namespace: instance.Namespace})
+	if err != nil {
+		return err
+	}
+	for _, installPlan := range installPlanList.Items {
+		Log.Info("Found installPlan", "name", installPlan.GetName())
+		// this should have a list containing the CSV names of all the old/legacy service operator CSVs
+		csvNames, found, err := uns.NestedSlice(installPlan.Object, "spec", "clusterServiceVersionNames")
+		if err != nil {
+			return err
+		}
+		if found {
+			// just checking for the first one should be sufficient
+			if isServiceOperatorResource(csvNames[0].(string)) {
+				err = r.Client.Delete(ctx, &installPlan)
+				if err != nil {
+					return err
+				}
+				Log.Info("Installplan deleted successfully", "name", installPlan.GetName())
+			}
+		}
+	}
+
+	return nil
+
+}
+
+// postCleanupObsoleteResources - deletes CSVs for old service operator bundles
+func (r *OpenStackReconciler) postCleanupObsoleteResources(ctx context.Context, instance *operatorv1beta1.OpenStack) error {
+	Log := r.GetLogger(ctx)
+
+	operatorGVR := schema.GroupVersionResource{
+		Group:    "operators.coreos.com",
+		Version:  "v1",
+		Resource: "operators",
+	}
+
+	// finally we can remove operator objects as all the refs have been cleaned up:
+	// 1) CSVs
+	// 2) Subscriptions
+	// 3) CRD olm.managed references removed
+	// 4) installPlan from old service operators removed
+	operatorList := &uns.UnstructuredList{}
+	operatorList.SetGroupVersionKind(operatorGVR.GroupVersion().WithKind("Operator"))
+	err := r.Client.List(ctx, operatorList, &client.ListOptions{Namespace: instance.Namespace})
+	if err != nil {
+		return err
+	}
+	for _, operator := range operatorList.Items {
+		Log.Info("Found Operator", "name", operator.GetName())
+		if isServiceOperatorResource(operator.GetName()) {
+
+			refs, found, err := uns.NestedSlice(operator.Object, "status", "components", "refs")
+			if err != nil {
+				return err
+			}
+			if found {
+
+				// The horizon-operator.openstack-operators has references to old roles/bindings
```

Comment:
Curious as to why Horizon is specifically referenced here.  Is this truly something unique to that operator?

Reply:
I took a look and didn't see anything jump out. It is just a Role and RoleBinding that cause the need for this code. All other objects seem to be cleaned up upon removal of the CSV/Subscriptions above.